metadata-ingestion/src/datahub/ingestion/source/powerbi.py

```diff
@@ -0,0 +1,1382 @@
+    ChangeAuditStamps,
+    DataPlatformInstance,
+)
+from datahub.metadata.schema_classes import (
+    BrowsePathsClass,
+    ChangeTypeClass,
+    ChartInfoClass,
+    ChartKeyClass,
+    CorpUserInfoClass,
+    CorpUserKeyClass,
+    DashboardInfoClass,
+    DashboardKeyClass,
+    DataPlatformInfoClass,
+    DatasetKeyClass,
+    DatasetPropertiesClass,
+    OwnerClass,
+    OwnershipClass,
+    OwnershipTypeClass,
+    StatusClass,
+)
+
+# Logger instance
+LOGGER = logging.getLogger(__name__)
+
+
+class Constant:
+    """
+    keys used in powerbi plugin
+    """
+
+    PBIAccessToken = "PBIAccessToken"
+    DASHBOARD_LIST = "DASHBOARD_LIST"
+    TILE_LIST = "TILE_LIST"
+    DATASET_GET = "DATASET_GET"
+    REPORT_GET = "REPORT_GET"
+    DATASOURCE_GET = "DATASOURCE_GET"
+    TILE_GET = "TILE_GET"
+    ENTITY_USER_LIST = "ENTITY_USER_LIST"
+    SCAN_CREATE = "SCAN_CREATE"
+    SCAN_GET = "SCAN_GET"
+    SCAN_RESULT_GET = "SCAN_RESULT_GET"
+    Authorization = "Authorization"
+    WorkspaceId = "WorkspaceId"
+    DashboardId = "DashboardId"
+    DatasetId = "DatasetId"
+    ReportId = "ReportId"
+    SCAN_ID = "ScanId"
+    Dataset_URN = "DatasetURN"
+    CHART_URN = "ChartURN"
+    CHART = "chart"
+    CORP_USER = "corpuser"
+    CORP_USER_INFO = "corpUserInfo"
+    CORP_USER_KEY = "corpUserKey"
+    CHART_INFO = "chartInfo"
+    STATUS = "status"
+    CHART_ID = "powerbi.linkedin.com/charts/{}"
+    CHART_KEY = "chartKey"
+    DASHBOARD_ID = "powerbi.linkedin.com/dashboards/{}"
+    DASHBOARD = "dashboard"
+    DASHBOARD_KEY = "dashboardKey"
+    OWNERSHIP = "ownership"
+    BROWSERPATH = "browsePaths"
+    DASHBOARD_INFO = "dashboardInfo"
+    DATAPLATFORM_INSTANCE = "dataPlatformInstance"
+    DATASET = "dataset"
+    DATASET_ID = "powerbi.linkedin.com/datasets/{}"
+    DATASET_KEY = "datasetKey"
+    DATASET_PROPERTIES = "datasetProperties"
+    VALUE = "value"
+    ENTITY = "ENTITY"
+    ID = "ID"
+
+
+class PowerBiAPIConfig(ConfigModel):
+    # Organsation Identifier
+    tenant_id: str
+    # PowerBi workspace identifier
+    workspace_id: str
+    # Current working environment. Possible values are DEV, QA, STAGE, PROD
+    # It is used in generating dataset URN
+    env: str
+    # Dataset type mapping
+    dataset_type_mapping: Dict[str, str]
+    # Azure app client identifier
+    client_id: str
+    # Azure app client secret
+    client_secret: str
+    # timeout for meta-data scanning
+    scan_timeout: int = 60
+
+    scope: str = "https://analysis.windows.net/powerbi/api/.default"
+    base_url: str = "https://api.powerbi.com/v1.0/myorg/groups"
+    admin_base_url = "https://api.powerbi.com/v1.0/myorg/admin"
+    authority = "https://login.microsoftonline.com/"
+
+    def get_authority_url(self):
+        return "{}{}".format(self.authority, self.tenant_id)
+
+
+class PowerBiDashboardSourceConfig(PowerBiAPIConfig):
+    platform_name: str = "powerbi"
+    platform_urn: str = "urn:li:dataPlatform:powerbi"
```

Comment:
use the utilities in `datahub.utilities` to construct urns

Reply:
Updated the code to use utility method